---
title: "Analysis Report 1: The relationship of human skin microbiome to host sex and environmental interactions"
author: "Emma Gibson"
date: "November 1, 2017"
output: github_document
bibliography: references.bib
csl: bioinformatics.csl
---

# Introduction

In recent years, the bacterium that live on and inside the human body have become a major subject of interest, as improved sequencing techniques make it easier than ever to study the previously difficult-to-observe microscopic world. Of the numerous microbial ecosystems that exist throughout the human body, few are as dynamic or diverse as that of the human skin. For example, even the microbiome on the left and right hands of the same person can have fairly few organisms in common. [@Caporaso2011] Because the skin microbiome is known to have an impact on and serve as an indicator of its host's health, it is important to study it and understand what its patterns reveal about its hosts. [@FREDRICKS2001167]

The skin microbiome's community composition can vary not only depending on what part of the body it is sampled from, but also from individual to individual. [@Costello1694] Although there is great variation in skin microbiome community composition from person to person, trends in microbiome communities can be found between people with various factors in common. For example, notable correlations between the microbial community composition have been found between people of the same handedness, hygiene habits, and sex. [@10.2307/25465398] Therefore, understanding and analyzing the patterns associated with the human skin microbiome could potentially allow one to infer information about its host.

The skin microbiome is not only unique to the skin of its host, however. It has been shown that a human leaves microbial evidence of their presence on the inanimate objects they interact with, tho the point where one can accurately identify what house belongs to what family based on microbial evidence. [@lax2014longitudinal] Understanding the patterns of peoples' skin microbiomes could lead to a greater comprehension of how human interactions affect their environment. That said, its usefulness in a field like forensics might be questionable due to the ever-changing nature of the human skin microbiome. [@Costello1694] The data I used in this analysis was originally collected in a paper by Fierer *et al.* in an attempt to use the skin microbiome for forensic purposes.However, I used this data to investigate what differences, if any, existed between the skin microbiomes of male and female individuals, as well as how that microbiome information could be tracked through the objects they have touched.

# Methods

## Sample origin and sequencing

Fierer *et al.* obtained the original samples from human volunteers and the keyboards/mice they touched. They collected them with sterilized cotton-tipped swabs that had been moistened by a sterile solution. They swabbed the hands midday, and swabbed the mice and keyboards within 12 hours of being used by their respective owners. All of the subjects they sampled were healthy, between 20-35 years old, and had not taken antibiotics 6 months before the sampling. They stored all swabs at -80 degrees Celsius for less than a week before DNA extraction.Fierer *et al.* obtained the bacterial gene sequences using high-throughput 454 pyrosequencing, one of several next-generation techniques. They sequenced the 16S ribosomal gene, a gene that is conserved across all of life and is conventionally used to identify bacterium.

## Computational

I processed and analyzed the sequence data using a combination of bash and R. First, I used bash to process the samples in order to determine their length and quality. Sequences of sufficient quality were then analyzed in R, using all of the relevant metadata and sequence information. R analyses were conducted through the usage of base R, as well as vegan, DADA2, and phyloseq. Vegan was used for various statistical analyses, such as to create rarefaction curves. Phyloseq was used to add various microbiome-specific functions to ggplot in order to better visualize the data [@mcmurdie2013].

DADA2 was used for inter-sequence analysis and cleaning, such as forming OTUs and resolving minor sequence differences. The DADA2 pipeline begins with filtering paired fastq files by trimming them to a specific length and removing sequences that are too short, then further filtering based on number of expected errors, quality score, and number of ambiguous bases [@callahan2016]. Next, it can remove chimeras, which are sequences that are mare up of two parent sequences that result from sequencing errors. Lastly, it merges paired forward and reverse reads after all the sequences have been cleaned and de-noised, so that DADA2 can be extremely strict and require exact overlap.

# Results

```{r load-libraries, message = FALSE}
# Be sure to install these packages before running this script
# They can be installed either with the intall.packages() function
# or with the 'Packages' pane in RStudio

# load general-use packages
library("dplyr")
library("tidyr")
library("knitr")
library("ggplot2")
library("vegan")
library("RColorBrewer")

# this package allows for the easy inclusion of literature citations in our Rmd
# more info here: https://github.com/crsh/citr
# and here:
# http://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html
library("citr")

# These are the primary packages well use to clean and analyze the data
# this package needs to be installed from bioconductor -- it's not on CRAN
# see info here: https://benjjneb.github.io/dada2/dada-installation.html
library("dada2")

# This to export a fasta of our final denoised sequence variants
library("seqinr")

# To install this you have to install from GitHub
# See more info here: https://github.com/leffj/mctoolsr
# run this -- install.packages("devtools")
# and then this -- devtools::install_github("leffj/mctoolsr")
library("mctoolsr")

# And this to visualize our results
# it also needs to be installed from bioconductor
library("phyloseq")
```

```{r extract-sample-and-file-names}
# NOTE: Much of the following follows the DADA2 tutorials available here:
# https://benjjneb.github.io/dada2/tutorial.html
# Accessed October 19, 2017

# set the base path for our input data files
path <- "data/raw_data"

# Sort ensures samples are in order
filenames_forward_reads <- sort(list.files(path, pattern = ".fastq"))

# Extract sample names, assuming filenames have format: SAMPLENAME.fastq
sample_names <- sapply(strsplit(filenames_forward_reads, "\\."), `[`, 1)

# Specify the full path to each of the filenames_forward_reads
filenames_forward_reads <- file.path(path, filenames_forward_reads)
```

```{r check-quality-plots}
# Plots the quality profiles of all twenty samples
plotQualityProfile(filenames_forward_reads[1:20])
```

We can see from the quality profiles that most reads tend to get pretty bad in quality after around 200 bases. Therefore, we decided to set a maximum acceptable sequence length of 225 bases.

```{r filter-reads}
# Place filtered files in filtered/ subdirectory
# note this will fail if the directory doesn't exist
filter_path <- file.path("output", "filtered")
filtered_reads_path <- file.path(filter_path,
                                 paste0(sample_names,
                                        "_filt.fastq.gz"))

# See ?filterAndTrim for details on the parameters
# See here for adjustments for 454 data:
# https://benjjneb.github.io/dada2/
#     faq.html#can-i-use-dada2-with-my-454-or-ion-torrent-data
filtered_output <- filterAndTrim(fwd = filenames_forward_reads,
                                 filt = filtered_reads_path,
                                 maxLen = 225,
                                 maxN = 0, # discard any seqs with Ns
                                 maxEE = 3, # allow w/ up to 3 expected errors
                                 truncQ = 2, # cut off if quality gets this low
                                 rm.phix = TRUE,
                                 compress = TRUE,
                                 multithread = FALSE)
```

```{r filtered-read-counts-table}
# produce nicely-formatted markdown table of read counts
# before/after trimming
kable(filtered_output,
      col.names = c("Reads In",
                    "Reads Out"))
```

```{r learn-errors}
# this build error models from each of the samples
errors_forward_reads <- learnErrors(filtered_reads_path,
                                    multithread = FALSE)
```

```{r visualize-errors-with-plots}
# quick check to see if error models match data
# (black lines match black points) and are generally decresing left to right
plotErrors(errors_forward_reads,
           nominalQ = TRUE)
```

```{r dereplicate-sequences}
# get rid of any duplicated sequences
dereplicated_forward_reads <- derepFastq(filtered_reads_path,
                                         verbose = TRUE)

# Name the derep-class objects by the sample names
names(dereplicated_forward_reads) <- sample_names
```

```{r run-dada}
# parameters adjusted based on recommendations for 454 data here:
# https://benjjneb.github.io/dada2/
#     faq.html#can-i-use-dada2-with-my-454-or-ion-torrent-data
dada_forward_reads <- dada(dereplicated_forward_reads,
                           err = errors_forward_reads,
                           HOMOPOLYMER_GAP_PENALTY = -1, # reduce penalty bc 454
                           BAND_SIZE = 32) # performs local alignments bc indels

# check dada results
dada_forward_reads
```

```{r make-sequence-table}
# produce the 'site by species matrix'
sequence_table <- makeSequenceTable(dada_forward_reads)
```

The output table has `r nrow(sequence_table)` rows (samples) and `r ncol(sequence_table)` columns (sequence variants). Notice how we can embed R code directly in our markdown text.

```{r histogram-of-sequence-lengths}
# Quick check to look at distribution of trimmed and denoised sequences
hist(nchar(getSequences(sequence_table)),
     main = "Histogram of fingal sequence variant lengths",
     xlab = "Sequence length in bp")
```

```{r remove-chimeras}
# Check for and remove chimeras
sequence_table_nochim <- removeBimeraDenovo(sequence_table,
                                            method = "consensus",
                                            multithread = FALSE,
                                            verbose = TRUE)

# What percent of our reads are non-chimeric?
non_chimeric_reads <- round(sum(sequence_table_nochim) / sum(sequence_table),
                            digits = 4) * 100
```

After removing chimeras, we were left with `r non_chimeric_reads`% of our cleaned reads.

```{r table-of-pipeline-read-counts}
# Build a table showing how many sequences remain at each step of the pipeline
get_n <- function(x) sum(getUniques(x)) # make a quick function
track <- cbind(filtered_output, # already has 2 columns
               sapply(dada_forward_reads, get_n),
               rowSums(sequence_table),
               rowSums(sequence_table_nochim))

# add nice meaningful column names
colnames(track) <- c("Input",
                     "Filtered",
                     "Denoised",
                     "Sequence Table",
                     "Non-chimeric")

# set the proper rownames
rownames(track) <- sample_names

# produce nice markdown table of progress through the pipeline
kable(track)
```

```{r assign-taxonomy}
# assigns taxonomy to each sequence variant based on a supplied training set
# made up of known sequences
taxa <- assignTaxonomy(sequence_table_nochim,
                       "data/training/rdp_train_set_16.fa.gz",
                       multithread = FALSE,
                       tryRC = TRUE) # also check with seq reverse compliments

# show the results of the taxonomy assignment
unname(taxa)
```

```{r extract-sequences-to-fasta}
# we want to export the cleaned, trimmed, filtered, denoised sequence variants
# so that we can build a phylogeny - we'll build the phylogeny outside of R
# but we need the fasta file to do so. We keep the names of each sequence as the
# sequence itself (which is rather confusing), because that's how DADA2 labels
# it's columns (e.g. 'species')
# function taken from https://github.com/benjjneb/dada2/issues/88
export_taxa_table_and_seqs <- function(sequence_table_nochim,
                                       file_seqtab,
                                       file_seqs) {
  seqtab_t <- as.data.frame(t(sequence_table_nochim)) # transpose to data frame
  seqs <- row.names(seqtab_t) # extract rownames
  row.names(seqtab_t) <- seqs # set rownames to sequences
  outlist <- list(data_loaded = seqtab_t)
  mctoolsr::export_taxa_table(outlist, file_seqtab) # write out an OTU table
  seqs <- as.list(seqs)
  seqinr::write.fasta(seqs, row.names(seqtab_t), file_seqs) # write out fasta
}

# actually run the function, with the names of the files we want it to create
# and where to put them
export_taxa_table_and_seqs(sequence_table_nochim,
                           "output/sequence_variants_table.txt",
                           "output/sequence_variants_seqs.fa")
```


```{r read-in-metadata-and-create-phyloseq}
# Next we want to read in the metadata file so we can add that in too
# This is not a csv file, so we have to use a slightly different syntax
# here the `sep = "\t"` tells the function that the data are tab-delimited
# and the `stringsAsFactors = FALSE` tells it not to assume that things are
# categorical variables
metadata_in <- read.table(paste0("data/metadata/",
                                 "fierer_forensic_hand_mouse_SraRunTable.txt"),
                          sep = "\t",
                          header = TRUE,
                          stringsAsFactors = FALSE,
                          row.names = 6) # sets sample IDs to row names

# read in the phylogeny, which was created from the fasta exported above
# in Geneious by aligning the sequences with MAFFT and then building a
# Maximum-Likelihood tree with RAxML
tree_in <- read_tree("output/sequence_variants_MAFFT_RAxML.newick")

# Construct phyloseq object (straightforward from dada2 outputs)
phyloseq_obj <- phyloseq(otu_table(sequence_table_nochim,
                                   taxa_are_rows = FALSE), # sample-spp matrix
                         sample_data(metadata_in), # metadata for each sample
                         tax_table(taxa), # taxonomy for each sequence variant
                         phy_tree(tree_in)) # phylogeny from sequence variants

melted_phyloseq_obj <- psmelt(phyloseq_obj)
```

```{r species-diversity}
# this code plots species diversity in male vs female hands
aggregated_sequence_data <- aggregate(sequence_table_nochim,
                            by = list(sex = metadata_in$sex_s),
                            FUN = sum)
row.names(aggregated_sequence_data) <- aggregated_sequence_data$sex
aggregated_sequence_data <- aggregated_sequence_data[
  2:length(aggregated_sequence_data)]
rarecurve(aggregated_sequence_data,
          main = "Species diversity in male vs female hands")
```

###Figure 1. Rarefaction curve of species diversity in male and female hands
This figure shows rarefaction curves for male hands, female hands, and mouse/keyboard samples. This type of analysis takes both sample size and sample diversity into account, and functions on the assumption that the more samples are taken from a specific community, the more diversity will be found.


```{r example-phyloseq-plot-1}
# this code looks at alpha diversity in male vs female hands
subsetted_pls_obj <- subset_samples(phyloseq_obj, sex_s != "Not applicable")
# this subset removes mouse data from the dataset
plot_richness(subsetted_pls_obj,
              x = "sex_s",
              measures = c("Shannon", "Simpson")) +
  xlab("Sample origin") +
  geom_boxplot(aes(fill = sex_s),
               width = 0.2) +
  theme(axis.text.x = element_text(angle = 0, hjust = 1)) +
  ggtitle("Alpha diversity in male vs female hands")
```

###Figure 2. Alpha diversity measures in each gender
This figure shows boxplots that represent the total diversity in both male and female hands, using both Shannon and Simpson diversity metrics. The line in the middle represents the median, the box represents the second and third quartiles, and the dots represent individual data points and outliers.

```{r phylum-by-sex}
# this code plots the abundance of each phylum present in each sex's hands
plot_bar(subsetted_pls_obj,
         x = "sex_s",
         y = "Abundance",
         fill = "Phylum") +
  ggtitle("Abundance of all phyla in each sex") +
  theme(axis.text.x = element_text(angle = 0, hjust = 1)) +
  scale_fill_brewer(palette = "Paired")
```

###Figure 3. Abundance of various phyla in each sex
This figure shows a stacked bar graph that represents the abundance of bacterial phyla on each hand, separated by sex.

```{r proteobacteria-in-hands-v-mice, fig.height=7}
# this code looks at the abundance of each phylum in each
# individual, and again pairs humans with their mice
ggplot(melted_phyloseq_obj,
        aes(x = host_subject_id_s,
            y = Abundance,
        color = sample_type_s)) +
  xlab("Individual") +
  geom_bar(aes(fill = sample_type_s),
        stat = "Identity",
        position = "dodge") +
  theme(axis.text.x = element_text(angle = 0, hjust = 1),
        strip.text.y = element_text(angle = 360)) +
  ggtitle("Phylum abundance in hands and mice") +
  facet_grid(Phylum ~ substr(melted_phyloseq_obj$host_subject_id_s, 1, 1),
             scales = "free_x")
```

###Figure 4. Abundance of each phylum in human hands and their respective mice
This figure shows the overall abundance of each phylum in each individual. Once again, it distinguished between each subject's hand and their respective mouse.

```{r male-female-hands-vs-mice}
# this code looks at the differences in diversity between
# male and female hands compared to the mice they used
plot_richness(phyloseq_obj,
              x = "host_subject_id_s",
              measures = c("Shannon"),
              color = "sample_type_s") +
  xlab("Individual") +
  geom_bar(aes(fill = sample_type_s),
        stat = "Identity",
        position = "dodge") +
  ggtitle("Shannon diversity in mice and hands") +
  theme(axis.text.x = element_text(angle = 0, hjust = 1)) +
# this next line tells R to only consider the first letter
# of the host ID (M2, F4, etc.), so the figure splits by sex
  facet_wrap(~ rep(c("F", "M"), each = 10),
             scales = "free_x")
```

###Figure 5. Diversity in human hands compared to their respective mice
This figure shows the Shannon diversity in each human hand, compared to the diversity on the respective mouse that said individual used.



```{r example-phyloseq-plot-2}
# this code makes a phylogenetic tree with points colored by individual ID
plot_tree(phyloseq_obj,
          color = "host_subject_id_s",
          ladderize = TRUE) +
# the last line arranges the tree branches from short to long
  ggtitle("A. Phylogenetic tree colored by host ID")
# this code makes a phylogenetic tree with points colored by host sex
plot_tree(phyloseq_obj,
          color = "sex_s",
          ladderize = TRUE) +
  ggtitle("B. Phylogenetic tree colored by host sex")
```

###Figure 6: Inferred phylogeny of sequences
This figure shows a phylogeny tree with points on tips representing samples within which each particular taxa occurred. These trees represent maximum likelihood phylogeny inferred using RAxML.6A shows the tree where each point's color represents the individual it came from, whereas 6B shows the same tree where the points' colors represent the sex of the host they came from.


# Discussion

In my analysis of this data, I focused on the interaction between gender and hand microbiome, as well as the relationship between one's hand microbiome and the inanimate objects they interact with.
For human subjects of either sex, and for mouse samples as well, I found that Fierer *et al.* appeared to have samples the full range of bacterial community diversity. Despite the varying sample sizes, every curve in the rarefaction curve forms a plateau after its initially steep incline, indicating that even if more samples were to be included, it is unlikely that the amount of diversity found would change significantly (fig 1). One possible explanation for this is that a larger sample size per individual could have lead to an over saturation of the most common phyla, therefore reducing the apparent diversity by comparison.

Despite the fact that there was a greater number of sequences for male test subjects, they showed less overall diversity than females, no matter which alpha diversity metrics are used (fig. 2). However, both male and female hands show wide variability in diversity between individuals, as shown by the outliers (fig. 2) In order to accurately determine whether the male or female hand microbiome is more diverse, a larger number of individuals would be needed. That said, in another study, Fierer *et al.* reached a similar conclusion about the larger diversity of the female hand microbiome in a sample size of 51 individuals. [@10.2307/25465398]

Of all the phyla present on human test subjects, the most abundant in both male and female hands was *proteobacteria* (fig. 3). However, it was vastly more common in male hands in terms of both overall abundance and proportion of the overall community makeup (fig. 3). This abundance of *proteobacteria* could be the reason that male hands show significantly lower diversity than female hands. That said, because *proteobacteria* is a massive and diverse phylum, it is hard to draw conclusions about what its abundance on human hands indicates about its hosts.

While *proteobacteria* is present in nearly every individual, most of the less-abundant phyla are only found in one or two of the individuals (fig. 4). That said, there are several phyla that have small abundance in each individual, but a larger abundance in their respective sex as a whole. For example, *actinobacteria* is the second most common phylum in female hands (fig. 3), but none of the female hands or mice have a particularly large abundance of it (fig. 4). This is unsurprising, as *actinobacteria* is a phylum of relatively common soil bacterium that one might expect to find in the environment, but would not necessarily thrive on a human hand. The *firmicutes* phylum shows a similar abundance pattern in both sexes, where it is present in small quantities on nearly every individual's hand (fig. 4). Like *proteobacteria*, *firmicutes* contains a wide variety of species, from pathogens to common gut symbiotes. The species found on the subjects' hands were likely from their gut microbiome, however, as *firmicutes* is one of the most common phyla in the human gut microbiome. [@LEY2006837] Therefore, it is unsurprising to find this phylum on human hands, as is known to exist in other parts of the body.

There does not appear to be a phylum that is prominently found on mice, but absent on human hands (fig. 4). However, the opposite is not true in all cases. For example, while the *bacteroidetes* phylum was found in great abundance on several female hands, neither of their respective mice nor any of the other mice show a great abundance of this phylum (fig. 4). Most individuals have patterns of *bacteriodetes* abundance comparable to *firmicutes*, with a small amount present in each individual. Again, this makes sense because *bacteriodetes*, together with *firmicutes*, make up around 98% of the mammalian gut microbiome. [@LEY2006837] However, two female individuals show a much larger abundance of it on their hands, but not on their respective mice. It makes sense that *bacteroidetes* was more common on female hands than male hands, given that it is known to be part of the placenta microbiome, and its prominence on these womens' hands could indicate several thins about their physiological state [@MOR2015S131]. For instance, they could be pregnant or in a different phase of their menstrual cycle from the other female test subjects. However, this would not explain why it was so abundant on their hands, yet absent on the computer mice they used.

There also appears to be a difference in the amount of diversity between male and female hands and the objects they have touched, as well. Whereas all but one female has more diversity on  her hand than on the mouse she touched, the opposite is true in males (fig. 5). While this could be due to the generally low diversity in male hands, this seems unlikely because the diversity on males' mice is not simply higher than the diversity of male hands, but also higher than the diversity in females' mice (fig 5). This could be due to the fact that *proteobacteria* is less abundant in male computer mice than it is in male hands (fig. 4). One possible explanation for why *proteobacteria* is more abundant on male hands than male mice is that the environment of the skin is favorable for their growth. If this is true, then the only *proteobacteria* found on mice would be there because it rubbed off from a human hand, but those on the human hand might be the progeny of the original bacteria that landed and proliferated there.

A phylogenetic tree of all of the obtained sequences shows that while distinct clades form, there is no apparent pattern between clades and host individuals. There appears to be no strong correlations of certain clades belonging primarily to one individual's hand and mouse biome (fig. 6A). While there may be a couple of close nodes that belong to the same individual, there are no large clades that correlate to any one individual. Every large clade appears to contain samples from both sexes as well, although there are several clades that do not contain the 'Not Applicable' sex given to mouse samples (fig. 6B). This indicates that these clades contain bacteria that is more specifically adapted to the human body, because they were found only on hands and not on mice.

# Sources Cited


