---
title: "Analysis Report 1: Your Title Here"
author: "Don Francisco"
date: "October 20, 2017"
output: github_document
bibliography: references.bib
csl: bioinformatics.csl
---

# Introduction

Add about 1.5-2 pages here. Must cite at least 5 peer reviewed articles.

# Methods

## Sample origin and sequencing

Fierer *et al.* obtained the riginal samples from human volunteers and the keyboards/mice they touched. They collected them with sterilized cotton-tipped swabs that had been moistened by a sterile solution. They swabbed the hands midday, and swabbed the mice and keyboards within 12 hours of being used by their respective owners. All of the subjects they sampled were healthy, between 20-35 years old, and had not taken antibiotics 6 months before the sampling. They stored all swabs at -80 degrees celcius for less than a week before DNA extraction.Fierer *et al.* obtained the becterial gene sequences using high-throughput 454 pyrosequencing, one of several next-generation techniques. They sequenced the 16S ribosomal gene, a gene that is conserved acros all of life and is conventionally used to identify bacterium.

## Computational

DSATDFHGJADJHGFDJGHAFDJGDSFAHGDFSAJHWRITE MORE ABOUT DADA2

I proscessed and analyzed the sequence data using a combination of bash and R. First, I used bash to proscess the samples in order to determine their length and quality. Sequences of sufficient quality were then analyzed in R, using all of the relevant metadata and sequence information. R analyses were conducted through the usage of base R, as well as vegan, DADA2, and phyloseq. Vegan was used for various statistical analyses, such as to create rarefaction curves. Phyloseq was used to add various microbiome-specific functions to ggplot in order to better visualize the data [@mcmurdie2013].

DADA2 was used for inter-sequence analysis, such as forming OTUs and resolving minor sequence differences. The DADA2 pipeline begins with filtering paired fastq files by trimming them to a specific length and removing sequences that are too short, then further filtering based on number of expected errors, quality score, and number of ambiguous bases [@callahan2016]. 

# Results

In addition to a minimum of 3-4 figures/tables (and associated captions), you should include sufficient text in this section to describe what your findings were. Remember that in the results section you just describe what you found, but you don't interpret it - that happens in the discussion.

```{r load-libraries, message = FALSE}
# Be sure to install these packages before running this script
# They can be installed either with the intall.packages() function
# or with the 'Packages' pane in RStudio

# load general-use packages
library("dplyr")
library("tidyr")
library("knitr")
library("ggplot2")
library("vegan")
library("RColorBrewer")

# this package allows for the easy inclusion of literature citations in our Rmd
# more info here: https://github.com/crsh/citr
# and here:
# http://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html
library("citr")

# These are the primary packages well use to clean and analyze the data
# this package needs to be installed from bioconductor -- it's not on CRAN
# see info here: https://benjjneb.github.io/dada2/dada-installation.html
library("dada2")

# This to export a fasta of our final denoised sequence variants
library("seqinr")

# To install this you have to install from GitHub
# See more info here: https://github.com/leffj/mctoolsr
# run this -- install.packages("devtools")
# and then this -- devtools::install_github("leffj/mctoolsr")
library("mctoolsr")

# And this to visualize our results
# it also needs to be installed from bioconductor
library("phyloseq")
```

```{r extract-sample-and-file-names}
# NOTE: Much of the following follows the DADA2 tutorials available here:
# https://benjjneb.github.io/dada2/tutorial.html
# Accessed October 19, 2017

# set the base path for our input data files
path <- "data/raw_data"

# Sort ensures samples are in order
filenames_forward_reads <- sort(list.files(path, pattern = ".fastq"))

# Extract sample names, assuming filenames have format: SAMPLENAME.fastq
sample_names <- sapply(strsplit(filenames_forward_reads, "\\."), `[`, 1)

# Specify the full path to each of the filenames_forward_reads
filenames_forward_reads <- file.path(path, filenames_forward_reads)
```

```{r check-quality-plots}
# Plots the quality profiles of all twenty samples
plotQualityProfile(filenames_forward_reads[1:20])
```

We can see from the quality profiles that most reads tend to get pretty bad in quality after around 200 bases. Therefore, we decided to set a maximum acceptable sequence length of 225 bases.

```{r filter-reads}
# Place filtered files in filtered/ subdirectory
# note this will fail if the directory doesn't exist
filter_path <- file.path("output", "filtered")
filtered_reads_path <- file.path(filter_path,
                                 paste0(sample_names,
                                        "_filt.fastq.gz"))

# See ?filterAndTrim for details on the parameters
# See here for adjustments for 454 data:
# https://benjjneb.github.io/dada2/
#     faq.html#can-i-use-dada2-with-my-454-or-ion-torrent-data
filtered_output <- filterAndTrim(fwd = filenames_forward_reads,
                                 filt = filtered_reads_path,
                                 maxLen = 225,
                                 maxN = 0, # discard any seqs with Ns
                                 maxEE = 3, # allow w/ up to 3 expected errors
                                 truncQ = 2, # cut off if quality gets this low
                                 rm.phix = TRUE,
                                 compress = TRUE,
                                 multithread = FALSE)
```

```{r filtered-read-counts-table}
# produce nicely-formatted markdown table of read counts
# before/after trimming
kable(filtered_output,
      col.names = c("Reads In",
                    "Reads Out"))
```

```{r learn-errors}
# this build error models from each of the samples
errors_forward_reads <- learnErrors(filtered_reads_path,
                                    multithread = FALSE)
```

```{r visualize-errors-with-plots}
# quick check to see if error models match data
# (black lines match black points) and are generally decresing left to right
plotErrors(errors_forward_reads,
           nominalQ = TRUE)
```

```{r dereplicate-sequences}
# get rid of any duplicated sequences
dereplicated_forward_reads <- derepFastq(filtered_reads_path,
                                         verbose = TRUE)

# Name the derep-class objects by the sample names
names(dereplicated_forward_reads) <- sample_names
```

```{r run-dada}
# parameters adjusted based on recommendations for 454 data here:
# https://benjjneb.github.io/dada2/
#     faq.html#can-i-use-dada2-with-my-454-or-ion-torrent-data
dada_forward_reads <- dada(dereplicated_forward_reads,
                           err = errors_forward_reads,
                           HOMOPOLYMER_GAP_PENALTY = -1, # reduce penalty bc 454
                           BAND_SIZE = 32) # performs local alignments bc indels

# check dada results
dada_forward_reads
```

```{r make-sequence-table}
# produce the 'site by species matrix'
sequence_table <- makeSequenceTable(dada_forward_reads)
```

The output table has `r nrow(sequence_table)` rows (samples) and `r ncol(sequence_table)` columns (sequence variants). Notice how we can embed R code directly in our markdown text.

```{r histogram-of-sequence-lengths}
# Quick check to look at distribution of trimmed and denoised sequences
hist(nchar(getSequences(sequence_table)),
     main = "Histogram of fingal sequence variant lengths",
     xlab = "Sequence length in bp")
```

```{r remove-chimeras}
# Check for and remove chimeras
sequence_table_nochim <- removeBimeraDenovo(sequence_table,
                                            method = "consensus",
                                            multithread = FALSE,
                                            verbose = TRUE)

# What percent of our reads are non-chimeric?
non_chimeric_reads <- round(sum(sequence_table_nochim) / sum(sequence_table),
                            digits = 4) * 100
```

After removing chimeras, we were left with `r non_chimeric_reads`% of our cleaned reads.

```{r table-of-pipeline-read-counts}
# Build a table showing how many sequences remain at each step of the pipeline
get_n <- function(x) sum(getUniques(x)) # make a quick function
track <- cbind(filtered_output, # already has 2 columns
               sapply(dada_forward_reads, get_n),
               rowSums(sequence_table),
               rowSums(sequence_table_nochim))

# add nice meaningful column names
colnames(track) <- c("Input",
                     "Filtered",
                     "Denoised",
                     "Sequence Table",
                     "Non-chimeric")

# set the proper rownames
rownames(track) <- sample_names

# produce nice markdown table of progress through the pipeline
kable(track)
```

```{r assign-taxonomy}
# assigns taxonomy to each sequence variant based on a supplied training set
# made up of known sequences
taxa <- assignTaxonomy(sequence_table_nochim,
                       "data/training/rdp_train_set_16.fa.gz",
                       multithread = FALSE,
                       tryRC = TRUE) # also check with seq reverse compliments

# show the results of the taxonomy assignment
unname(taxa)
```

```{r extract-sequences-to-fasta}
# we want to export the cleaned, trimmed, filtered, denoised sequence variants
# so that we can build a phylogeny - we'll build the phylogeny outside of R
# but we need the fasta file to do so. We keep the names of each sequence as the
# sequence itself (which is rather confusing), because that's how DADA2 labels
# it's columns (e.g. 'species')
# function taken from https://github.com/benjjneb/dada2/issues/88
export_taxa_table_and_seqs <- function(sequence_table_nochim,
                                       file_seqtab,
                                       file_seqs) {
  seqtab_t <- as.data.frame(t(sequence_table_nochim)) # transpose to data frame
  seqs <- row.names(seqtab_t) # extract rownames
  row.names(seqtab_t) <- seqs # set rownames to sequences
  outlist <- list(data_loaded = seqtab_t)
  mctoolsr::export_taxa_table(outlist, file_seqtab) # write out an OTU table
  seqs <- as.list(seqs)
  seqinr::write.fasta(seqs, row.names(seqtab_t), file_seqs) # write out fasta
}

# actually run the function, with the names of the files we want it to create
# and where to put them
export_taxa_table_and_seqs(sequence_table_nochim,
                           "output/sequence_variants_table.txt",
                           "output/sequence_variants_seqs.fa")
```


```{r read-in-metadata-and-create-phyloseq}
# Next we want to read in the metadata file so we can add that in too
# This is not a csv file, so we have to use a slightly different syntax
# here the `sep = "\t"` tells the function that the data are tab-delimited
# and the `stringsAsFactors = FALSE` tells it not to assume that things are
# categorical variables
metadata_in <- read.table(paste0("data/metadata/",
                                 "fierer_forensic_hand_mouse_SraRunTable.txt"),
                          sep = "\t",
                          header = TRUE,
                          stringsAsFactors = FALSE,
                          row.names = 6) # sets sample IDs to row names

# read in the phylogeny, which was created from the fasta exported above
# in Geneious by aligning the sequences with MAFFT and then building a
# Maximum-Likelihood tree with RAxML
tree_in <- read_tree("output/sequence_variants_MAFFT_RAxML.newick")

# Construct phyloseq object (straightforward from dada2 outputs)
phyloseq_obj <- phyloseq(otu_table(sequence_table_nochim,
                                   taxa_are_rows = FALSE), # sample-spp matrix
                         sample_data(metadata_in), # metadata for each sample
                         tax_table(taxa), # taxonomy for each sequence variant
                         phy_tree(tree_in)) # phylogeny from sequence variants

melted_phyloseq_obj <- psmelt(phyloseq_obj)
```

```{r species-diversity}
# this code plots species diversity in male vs female hands
aggregated_sequence_data <- aggregate(sequence_table_nochim,
                            by = list(sex = metadata_in$sex_s),
                            FUN = sum)
row.names(aggregated_sequence_data) <- aggregated_sequence_data$sex
aggregated_sequence_data <- aggregated_sequence_data[2:length(aggregated_sequence_data)]
rarecurve(aggregated_sequence_data,
          main = "Species diversity in male vs female hands")
```

###Figure 1. Rarefaction curve of species diversity in male and female hands
This figure shows rarefaction curves for male hands, female hands, and mouse/keyboard samples. This type of analysis takes both sample size and sample diversity into account, and functions on the assumption that the more samples are taken from a specific community, the more diversity will be found.


```{r example-phyloseq-plot-1}
# this code looks at alpha diversity in male vs female hands
subsetted_pls_obj <- subset_samples(phyloseq_obj, sex_s != "Not applicable")
# this subset removes mouse data from the dataset
plot_richness(subsetted_pls_obj,
              x = "sex_s",
              measures = c("Shannon", "Simpson")) +
  xlab("Sample origin") +
  geom_boxplot(aes(fill = sex_s),
               width = 0.2) +
  theme(axis.text.x = element_text(angle = 0, hjust = 1)) +
  ggtitle("Alpha diversity in male vs female hands")
```

###Figure 2. Alpha diversity measures in each gender
This figure shows boxplots that represent the total diversity in both male and femalle hands, using both Shannon and Simpson diversity metrics.

```{r phylum-by-sex}
# this code plots the abundance of each phylum present in each sex's hands
plot_bar(subsetted_pls_obj,
         x = "sex_s",
         y = "Abundance",
         fill = "Phylum") +
  ggtitle("Abundance of all phylums in each sex") +
  theme(axis.text.x = element_text(angle = 0, hjust = 1)) +
  scale_fill_brewer(palette = "Paired")
```

###Figure 3. Abundance of various phylums in each sex
This figure shows a bar graph that represents the relative abundance of bacterial phylums in each hand, sepparated by sex.

```{r male-female-hands-vs-mice}
# this code looks at the differences in diversity between
# male and female hands compared to the mice they used
plot_richness(phyloseq_obj,
              x = "host_subject_id_s",
              measures = c("Shannon"),
              color = "sample_type_s") +
  xlab("Individual") +
  geom_bar(aes(fill = sample_type_s), 
        stat = "Identity", 
        position = "dodge") +
  ggtitle("Shannon diversity in mice and hands") +
  theme(axis.text.x = element_text(angle = 0, hjust = 1)) +
# this next line tells R to only consider the first letter
# of the host ID (M2, F4, etc.), so the figure splits by sex
  facet_wrap(~ substr(metadata_in$host_subject_id_s, 1, 1),
             scales = "free_x")
```

###Figure 4. Diversity in human hands compared to their respective mice
This figure shows the Shannon diversity in each human hand, compared to the respective mouse that individual used.

```{r proteobacteria-in-hands-v-mice, fig.height=7}
# this code looks at the abundance of each phylum in each
# individual, and again pairs humans with their mice
ggplot(melted_phyloseq_obj,
        aes(x = host_subject_id_s,
            y = Abundance,
        color = sample_type_s)) +
  xlab("Individual") +
  geom_bar(aes(fill = sample_type_s), 
        stat = "Identity", 
        position = "dodge") +
  theme(axis.text.x = element_text(angle = 0, hjust = 1),
        strip.text.y = element_text(angle = 360)) +
  ggtitle("Phylum abundance in hands and mice") +
  facet_grid(Phylum ~ substr(melted_phyloseq_obj$host_subject_id_s, 1, 1), 
             scales = "free_x")
```

###Figure 5. Abundance of each phylum in human hands and their respective mice
This figure shows the overall abundance of each phylum in each individual. Once again, it distinguished between each subject's hand and their respective mouse.

```{r example-phyloseq-plot-2}
# phylogeny, yay!
plot_tree(phyloseq_obj,
          color = "sex_s",
          ladderize = TRUE) # this arranges the tree branches from short to long
```

###Figure 2: Inferred phylogeny of sequences
This figure shows a phylogenetic tree with points on tips representing samples within which each particular taxa occurred. This tree represents maximum likelihood phylogeny inferred using RAxML.


# Discussion

Add around 2-3 pages interpreting your results and considering future directions one might take in analyzing these data.

# Sources Cited


